src/lsm/cache_map.zig

```diff
@@ -203,73 +194,96 @@ pub fn CacheMapType(
         //       to the scope rollback log.
         //    b. If the item doesn't exist in the stash, it was an insert. Append a tombstone to
         //       the scope rollback log.
-        fn upsert_scope_opened(self: *Self, value: *const Value) void {
+        fn upsert_scope_opened(self: *CacheMap, value: *const Value) void {
             assert(self.scope_is_active);
 
-            const result = self.cache.upsert(value);
-
-            if (result.evicted) |evicted| {
-                switch (result.updated) {
-                    .update => {
-                        // Case 1.
-                        self.scope_rollback_log.appendAssumeCapacity(evicted);
-                    },
-                    .insert => {
-                        // Case 2.
-                        const gop = self.stash_1.getOrPutAssumeCapacity(evicted);
-                        gop.key_ptr.* = evicted;
-
-                        // Case 3 below handles appending into the rollback log if needed.
-                    },
+            const key = key_from_value(value);
+            if (self.cache) |*cache| {
```

Comment:
Not a fan of duplication all over the place. I wonder, would it be beneficial to introduce

```
fn cache_get(cache_map: *CacheMap)
fn cache_upsert(cache_map: *CacheMap)
```

which encapsulate null checking?

I _think_ null cache behaves as infinitely-sized cache --- it never evicts anything, and always reports that a new value was inserted? 

Ie, it's hard to verify at the glance that the logic in two branches are compatible here:


```
            if (self.cache) |*cache| {
                const maybe_removed = cache.remove(key);

                if (self.scope_is_active) {
                    // TODO: Actually, does the fuzz catch this...
                    self.scope_rollback_log.appendAssumeCapacity(
                        maybe_removed orelse
                            self.stash.getKey(tombstone_from_key(key)) orelse return,
                    );
                }

                // We always need to try remove from the stash; since it could have a stale value.
                // The early return above is OK - if it doesn't exist, there's nothing to remove.
                _ = self.stash.remove(tombstone_from_key(key));
            } else {
                const maybe_removed = self.stash.fetchRemove(tombstone_from_key(key));
                if (self.scope_is_active) {
                    if (maybe_removed) |kv| {
                        // TODO: Actually, does the fuzz catch this...
                        self.scope_rollback_log.appendAssumeCapacity(kv.key);
                    }
                }
            }

```

Reply:
I simplified the code by decoupling the upsert from the rollback logic.